### PR TITLE
Add cli command to count various types of gates

### LINKED ISSUE #343: Add gate count functionality
Currently there is no way to see the exact gate count for a particular circuit.

We can add a nargo command which tells us how many of each opcode we are using. 

Example :

```rust
Arith : 10
Sha256 : 200
CheckMembership: 4
```

### PR DESCRIPTION
**Pull request by TomAFrench**
closes #343 

This PR adds the `nargo gates` command which counts the number of each gate in the circuit. I've reused the existing names for the gadget calls and added names in a similar style for the "standard" gates.

Running this command on [mastermind-noir](https://github.com/vezenovm/mastermind-noir/tree/43516b105294162522d4de24ebb7a31192fa99db) gives the output below.
```
$ nargo gates
Gates successfully counted

Total gates: 336

By type:
arithmetic: 269
invert: 43
odd_range: 1
pedersen: 1
range: 13
split: 8
truncate: 1
```

Open to any opinions on formatting, etc.

### CODE REVIEW

File: crates/acir/src/circuit/gate.rs [hunk 1]
```diff
@@ -31,6 +31,20 @@ pub enum Gate {
...
 
 impl Gate {
+    pub fn name(&self) -> &str {
+        match self {
+            Gate::Arithmetic(_) => "arithmetic",
+            Gate::Range(_, _) => "range",
+            Gate::And(_) => "and",
+            Gate::Xor(_) => "xor",
+            Gate::Directive(Directive::Invert { x: _, result: _ }) => "invert",
+            Gate::Directive(Directive::Truncate { a: _, b: _, c: _, bit_size: _ }) => "truncate",
```
Comment (jfecher) [inline crates/acir/src/circuit/gate.rs:41]: To ignore all (or a subset of) fields in a struct pattern you can do `Directive::Invert { .. }`

**Review by jfecher — APPROVED** [review]
👍

File: crates/acir/src/circuit/gate.rs (hunk 1 above)
Comment (TomAFrench) [reply crates/acir/src/circuit/gate.rs]: Cheers, I hadn't seen that before.

**Review by vezenovm — APPROVED** [review]

**Comment by TomAFrench** [comment]
Let's merge this after #371 so it can make use of `compile_circuit`.

**Comment by vezenovm** [comment]
> 

#371 is merged in now

**Comment by vezenovm** [comment]
> Let's merge this after #371 so it can make use of `compile_circuit`.

PR #371 is merged in now

**Pull request merged**